File: content/examples/other/leetcode/problems/course-schedule/example-1/Solution.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

public class Solution {
  static boolean canFinishBFSKahn(int numCourses, int[][] prerequisites) {
    int[] indegree = new int[numCourses];
    List<List<Integer>> adj = new ArrayList<>();
    for (int i = 0; i < numCourses; i += 1) adj.add(new ArrayList<>());
    for (int[] p : prerequisites) {
      int course = p[0];
      int prereq = p[1];
      adj.get(prereq).add(course);
      indegree[course] += 1;
    }

    Deque<Integer> q = new ArrayDeque<>();
    for (int i = 0; i < numCourses; i += 1) if (indegree[i] == 0) q.addLast(i);

    int taken = 0;
    while (!q.isEmpty()) {
      int cur = q.removeFirst();
      taken += 1;
      for (int nxt : adj.get(cur)) {
        indegree[nxt] -= 1;
        if (indegree[nxt] == 0) q.addLast(nxt);
      }
    }
    return taken == numCourses;
  }

  public static void main(String[] args) {
    System.out.println(canFinishBFSKahn(2, new int[][] {{1, 0}}));
    System.out.println(canFinishBFSKahn(2, new int[][] {{1, 0}, {0, 1}}));
  }
}

File: content/examples/other/leetcode/problems/course-schedule/example-1/solution.py
from collections import deque


def can_finish_bfs_kahn(num_courses, prerequisites):
    indegree = [0] * num_courses
    adj = [[] for _ in range(num_courses)]
    for course, prereq in prerequisites:
        adj[prereq].append(course)
        indegree[course] += 1

    q = deque([i for i in range(num_courses) if indegree[i] == 0])
    taken = 0
    while q:
        cur = q.popleft()
        taken += 1
        for nxt in adj[cur]:
            indegree[nxt] -= 1
            if indegree[nxt] == 0:
                q.append(nxt)

    return taken == num_courses


if __name__ == "__main__":
    print(can_finish_bfs_kahn(2, [(1, 0)]))
    print(can_finish_bfs_kahn(2, [(1, 0), (0, 1)]))
